Add support for TorchScript tracing (#262)
explosion/curated-transformers: docs/source/deployment.rst

Deployment
==========

In many cases, a PyTorch model can be deployed directly in a Python application,
eg. in a `Flask-based REST service
<https://pytorch.org/tutorials/intermediate/flask_rest_api_tutorial.html>`_. In
other cases, such as deployment to non-CUDA accelerators, additional model
transformations might be needed. On this page, we cover several deployment
scenarios.

> TorchScript tracing

Reviewer: ```suggestion
TorchScript Tracing
```

-------------------

Many deployment methods start from `TorchScript`_. For instance, ONNX conversion
converts the TorchScript representation of a model. TorchScript is a
statically-typed subset of Python. It only supports the types that are necessary
for representing neural network models.

Curated Transformers supports TorchScript through `tracing`_.
Tracing runs the model with some example inputs and records the computation
graph. The TorchScript code is then generated from this computation graph,
discarding all other Python code.

.. _TorchScript: https://pytorch.org/docs/stable/jit.html
.. _tracing: https://pytorch.org/docs/stable/generated/torch.jit.trace.html#torch.jit.trace

> Tracing a model

Reviewer: ```suggestion
Tracing a Model
```

^^^^^^^^^^^^^^^

Models are traced using the |torch.jit.trace|_ function. The first argument to
this function is the model that you would like to trace, the second argument the
inputs as a tuple. For example, we can trace a decoder as follows:

.. code-block:: python

   import torch
   import torch.jit
   from curated_transformers.models import AutoDecoder

   device = torch.device("cuda", index=0)
   decoder = AutoDecoder.from_hf_hub(name="tiiuae/falcon-7b", device=device)
   X_example = torch.zeros(4, 20, dtype=torch.long, device=device)
   traced = torch.jit.trace(decoder, (X_example),))

As you can see, we are feeding the model with an all-zeros tensor during
tracing. This is not really an issue, as long as the inputs allows the model to
run normally, tracing can do its work.

In the example above, ``traced`` is a TorchScript module. From the surface, it
behaves like any other module. We can feed it some piece identifiers to get
their hidden representations:

.. code-block:: python

   from curated_transformers.tokenizers import AutoTokenizer

   tokenizer = AutoTokenizer.from_hf_hub(name="tiiuae/falcon-7b")
   pieces = tokenizer(["Hello world!", "This is a test"])

>    Y = traced(pieces.padded_tensor(padding_id=0).to(device))

Reviewer: ```suggestion
   Y = traced(pieces.padded_tensor(padding_id=0, device=device))
```

   assert isinstance(Y, tuple)
   last_layer = Y[0][-1]

> The model works as before, with one catch. Normally a decoder returns a

Reviewer: ```suggestion
The model works as before, albeit with one catch. Normally a decoder returns a
```

:py:class:`~curated_transformers.models.output.ModelOutputWithCache` instance,

> whereas the traced model returns a tuple. The reason is that TorchScript only

Reviewer: ```suggestion
but the traced model returns a tuple instead. The reason is that TorchScript only
```

supports a limited set of types. Since arbitrary types are not supported, we
convert the :py:class:`~curated_transformers.models.output.ModelOutputWithCache`

> type to a tuple in a traced model. The tuple will have the same ordering as the

Reviewer: ```suggestion
instance to a tuple in a traced model. The tuple will have the same ordering as the
```
